docs/add_a_validation.md

```diff
@@ -0,0 +1,295 @@
+                    break;
+                default:
+                    $validationErrors[] = new ValidationError("Validator parameter '$name' is unknown");
+            }
+        }
+    }
+
+    return $validationErrors;
+}
+```
+
+`validate()` is the method that runs the actual validation on data:
+
+``` php
+// eZ/Publish/FieldType/Tweet/Type.php
+
+public function validate(FieldDefinition $fieldDefinition, SPIValue $fieldValue)
+{
+    $errors = [];
+
+    if ($this->isEmptyValue($fieldValue)) {
+        return $errors;
+    }
+
+    // Tweet Url validation
+    if (!preg_match('#^https?://twitter.com/([^/]+)/status/[0-9]+$#', $fieldValue->url, $m)) {
+        $errors[] = new ValidationError(
+            'Invalid twitter status url %url%',
+            null,
+            ['%url%' => $fieldValue->url]
+        );
+
+        return $errors;
+    }
+
+    $author = $m[1];
+    $validatorConfiguration = $fieldDefinition->getValidatorConfiguration();
+    if (!$this->isAuthorApproved($author, $validatorConfiguration)) {
+        $errors[] = new ValidationError(
+            'Twitter user %user% is not in the approved author list',
+            null,
+            ['%user%' => $m[1]]
+        );
+    }
+
+    return $errors;
+}
+
+private function isAuthorApproved($author, $validatorConfiguration)
+{
+    return !isset($validatorConfiguration['TweetValueValidator'])
+        || empty($validatorConfiguration['TweetValueValidator']['authorList'])
+        || in_array($author, $validatorConfiguration['TweetValueValidator']['authorList']);
+}
+```
+
+In addition to validating the URL with a regular expression, which we implemented before, if your Field Type instance’s configuration contains a `TweetValueValidator` key, you will check that the username in the status URL matches one of the valid authors.
+
+## Implement FormMapper
+
+Next, we need to create a way for user to input a list of authors for our validation. To achieve this, we need to implement a FormMapper that allows us to add an input field to form with content field definition.
+The most basic one looks like the following:
+
+```php
+// eZ/Publish/FieldType/Tweet/FormMapper.php
+
+class FormMapper implements FieldDefinitionFormMapperInterface
+{
+    public function mapFieldDefinitionForm(FormInterface $fieldDefinitionForm, FieldDefinitionData $data)
+    {
+        $fieldDefinitionForm
+            ->add('authorList', TextType::class, [
+                'required' => false,
+                'property_path' => 'validatorConfiguration[TweetValueValidator][authorList]',
+                'label' => 'field_definition.eztweet.authorList'
+            ]);
+    }
+}
+```
+
+Unfortunately, because in our case authorList validator has array type, we cannot leave it like that. We will need a way to transform data from array to comma-separated string and vice versa.
+To achieve this, first we need to implement a DataTransformer:
+
+```php
+// Form/StringToArrayTransformer.php
+
```

Comment:
Wouldn't it be better to put the full file here? An over-eager copy-paster (like me) could overlook the first 15 lines and get errors (unless you consider it to be an obvious step for someone who uses this tutorial)